### DISCUSSION TITLE
How to properly set-up your MSYS2

### DISCUSSION THREAD

**Discussion by RickBarretto** [can-lehmann/owlkettle#144] [Show and tell]
## Installation

The first step is to install the MSYS2 into your machine, for this, see their [getting started][msys-start] and [installation instructions][msys-install].

Make sure you understood how they properly works because they provides [a lot of different toolchains][msys-envs] to work with. Read their [introduction][msys-intro] for this.

Once you understood their processes, concepts and installed it, you can go to the next step: Set up the environment.

[msys-start]: https://www.msys2.org/
[msys-install]: https://www.msys2.org/wiki/MSYS2-installation/
[msys-envs]: https://www.msys2.org/docs/environments/
[msys-intro]: https://www.msys2.org/wiki/MSYS2-introduction/

## Setting up your Mingw environment

In my example, I'll set up the MinGW64 environment with:

```bash
# $ pacman --sync --refresh --sysupgrade git
$ pacman --sync --refresh --sysupgrade base-devel
$ pacman --sync --refresh --sysupgrade mingw-w64-x86_64-toolchain
```

[`base-devel`][base-devel]: provides a set for building packages with makepkg.
[`mingw-w64-x86_64-toolchain`][min-toolchain]: provides a set of packages for development as GCC, Make, pkg-config, binutils...

> [!WARNING]
> I had some issues using the `git` from MSYS2 on VScode's extensions as GitLens. I recommend download the git from their [website][git-for-win] instead. The `git` itself works fine, but not when used from extensions.
> Or download from winget:
> ```
>  winget install --id Git.Git -e --source winget
> ```

[base-devel]: https://packages.msys2.org/base/base-devel
[min-toolchain]: https://packages.msys2.org/groups/mingw-w64-x86_64-toolchain
[git-for-win]: https://git-scm.com/download/win

## Setting up your system's PATH

After installed, make sure you added your paths into your system's PATH. This would be something similar to, in this order:

```
C:\Users\<you>\AppData\Local\Microsoft\WindowsApps
C:\Users\<you>\AppData\Local\Programs\Git\cmd
C:\Users\<you>\AppData\Local\Programs\Microsoft VS Code\bin
C:\Users\<you>\Downloads\nim-2.0.0_x64\nim-2.0.0\bin
C:\msys64\mingw64\bin
C:\msys64\mingw64\include
C:\msys64\mingw64\lib
C:\msys64\usr\bin
C:\msys64\usr\include
C:\msys64\usr\lib
```

I prefer put `mingw/bin` before the `usr/bin` to prefer the tools from MinGw64's environment instead of the MSYS environment. 
Notice that my `Git` was put before my msys64 ones because if you have two `git`s installed, it'll give preference to your git-for-windows.

If you don't know how to open you system's PATH, search for "Environment Variables" in your Windows' menu, then open the "Path" option and edit it.

**Comment by michaelsbradleyjr** [comment]
Per discussion in Discord earlier, some feedback re: MSYS2

1. It's good to consistently refer to it as MSYS2 and *not* MSYS, because they're completely separate projects and people do get confused about that kind of thing when they're doing web searches, etc. *(maybe they should have named it something else like MSYS-NG, whatever, but that ship sailed a long time ago)*

https://www.msys2.org/wiki/History/
> **MSYS2** is an independent rewrite of MSYS, based on modern Cygwin and MinGW-w64 with the aim of better interoperability with native Windows software... MinGW is a software distribution and development platform. It is accompanied by MSYS, an old Cygwin fork. The MSYS2 and MinGW-w64 projects are not associated with MSYS and MinGW, other than by the name and common goals. MSYS2 is ideologically a successor to MSYS and MinGW. MSYS -- although definitely useful -- is really old and getting in the way of developers. MSYS2 was created to replace the original MSYS while avoiding its problems.

---

2. The basic recommendation for environment–toolchain is to use UCRT64

https://www.msys2.org/docs/environments/
> The differences among the environments are mainly environment variables, default compilers/linkers, architecture, system libraries used etc. If you are unsure, go with UCRT64.

At one point MINGW64 was the basic recommendation, but it uses MSVCRT and at this point most projects don't need it unless it's a strict requirement. Also, [Microsoft's UCRT](https://learn.microsoft.com/en-us/cpp/porting/upgrade-your-code-to-the-universal-crt?view=msvc-170) has some clear advantages, e.g. proper support for the UTF-8 locale.

So, in the README, etc. you'd want to suggest
```
pacman -S base-devel git mingw-w64-ucrt-x86_64-toolchain
```
i.e. the `w64-ucrt-x86_64` packages instead of the `w64-x86_64` packages.

UCRT64 works great for Nim, but the best environment for Nim on x86_64 might be CLANG64. It has the advantages of UCRT64 over MINGW64 *and* it has native thread-local storage, not the emulated gcc stuff in UCRT64 and MINGW64, which can have a serious performance impact.
```
pacman -S base-devel git mingw-w64-clang-x86_64-toolchain
```

---

3. I don't think it's a good idea to try to mix MSYS2 paths into the Windows system path

MSYS2 toolchains are intended to be used in an MSYS2 Bash session *not* `cmd.exe` or PowerShell. Software *produced* with an MSYS2 toolchain can be completely independent of that toolchain, i.e. with static linking you should be able to build on Windows 10/11 in CLANG64 and ship a binary to an old Windows 7 x64 machine that has nothing of MSYS2 installed and it would run correctly in `cmd.exe`. But in development, you should stick to using MSYS2 Bash with the `PATH` setup it provides for the selected environment: CLANG64, UCRT64, et al.

**Reply by RickBarretto** [reply]
Thanks a lot for the advise and feedback! 🙂
I've never had used UCRT or CLANG64, I'll have a look on it.

**Reply by PimpNamedShimp** [reply]
Well... it's really ok if you just temporarily edit PATH to include /msys64/usr/bin just long enough to run a command or two, so you don't have to open another terminal/environment and then have to copy it to the mounted windows directory you actually wanted it on. Git already causes some performance issues with or within wsl, or it has as of late for some reason. Plus, if you have ever downloaded the windows git install package, when reading through install prompts, they actually recommend users extend some caution when using Git. I imagine, also, that having multiple environments alongside msys2, alone, could cause registry errors, configuration problems,... Especially, if your environments are not containerized or chrooted or whatever...
